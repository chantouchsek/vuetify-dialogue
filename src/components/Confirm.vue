<template>
  <v-card tile>
    <v-toolbar
      v-if="title"
      :dark="Boolean(getColor)"
      :color="getColor"
      dense
      flat
    >
      <v-icon
        v-if="Boolean(getIcon)"
        left
      >
        {{ getIcon }}
      </v-icon>
      <v-toolbar-title>{{ getTitle }}</v-toolbar-title>
    </v-toolbar>
    <v-card-text
      class="body-1 py-2"
      :class="{ 'pt-4': !title }"
      v-html="getText"
    />
    <dialog-actions
      :actions="actions"
      v-bind="getDialogOptions"
    />
  </v-card>
</template>

<script>
import Confirmable from 'vuedl/src/mixins/confirmable'
import Colorable from '../mixins/colorable'
import Iconable from '../mixins/iconable'
import Translatable from '../mixins/translatable'
import DialogActions from './DialogActions.vue'
import { VCard, VCardText, VToolbar, VToolbarTitle, VIcon } from 'vuetify/lib'
import defu from 'defu'

export default {
  components: {
    DialogActions,
    VCard,
    VCardText,
    VToolbar,
    VToolbarTitle,
    VIcon
  },
  layout: ['default', { width: 450 }],
  mixins: [Iconable, Confirmable, Colorable, Translatable],
  props: {
    actionOptions: Object
  },
  computed: {
    getDialogOptions () {
      return defu({ i18n: this.i18n }, ...this.actionOptions)
    },
    getText () {
      if (typeof this.text !== 'function') {
        return this.getTranslate(this.text)
      }
      return typeof this.text === 'function' ? this.text() : this.text
    },
    getTitle () {
      return this.getTranslate(this.title)
    }
  }
}
</script>
