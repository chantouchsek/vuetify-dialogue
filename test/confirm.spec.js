import Dialog from 'vuedl/src/dialog'
import DialogManager from 'vuedl/src/manager'
import Confirm from '../src/components/Confirm'
import Returnable from 'vuedl/src/mixins/returnable'
import Vue from 'vue'
import Vuetify from 'vuetify/lib'
Vue.use(Vuetify)
const vuetify = new Vuetify({})

describe('manager', () => {
  let manager
  test('Create manager instance', () => {
    manager = new DialogManager({ context: { store: {}, vuetify } })
    manager.component('confirm', Confirm)
  })

  test('Test default confirm', async () => {
    const res = await manager.confirm({
      text: 'Test confirmation',
      title: 'Title'
    })
    expect(res.vm.$el).toMatchSnapshot()
    await res.close()
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })

  test('Test confirm without icon', async () => {
    const res = await manager.confirm({
      icon: false,
      text: 'Test confirmation withot icon',
      title: 'Title'
    })
    expect(res.vm.$el).toMatchSnapshot()
    await res.close()
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })

  test('Returnable', async () => {
    const dlg = new Dialog({
      mixins: [Returnable],
      template: '<p></p>'
    })
    await dlg.show()
    expect(typeof dlg.vmd.return).toBe('function')
    expect(dlg.element).toMatchSnapshot()
    setTimeout(() => {
      dlg.vmd.return(123)
    }, 5)
    const res = await dlg.wait()
    expect(res).toBe(123)
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })

  test('Check confirm', async () => {
    const dlg = await manager.confirm({
      text: 'test',
      type: 'warning',
      actions: ['ok', 'cancel']
    })
    expect(dlg.vm.$el).toMatchSnapshot()
    setTimeout(() => {
      dlg.element.querySelector('[action-key=ok]').click()
    }, 0)
    const res = await dlg.wait()
    expect(res).toBe('ok')
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })

  test('Check confirm with action icons', async () => {
    const dlg = await manager.confirm({
      text: 'test',
      icon: 'warning',
      actions: {
        false: {
          icon: {
            text: 'cancel'
          },
          color: 'info',
          rounded: true,
          text: 'No'
        },
        true: {
          text: 'Yes',
          icon: {
            text: 'success',
            right: true
          }
        }
      }
    })
    expect(dlg.vm.$el).toMatchSnapshot()
    await dlg.close()
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })

  test('Check confirm with btns true|false', async () => {
    const dlg = await manager.confirm({
      text: 'test',
      actions: { true: 'Yes', false: 'No' }
    })
    expect(dlg.element).toMatchSnapshot()
    expect(dlg.element.querySelector('[action-key=true]')).toBeTruthy()
    expect(dlg.element.querySelector('[action-key=false]')).toBeTruthy()
    setTimeout(() => {
      dlg.element.querySelector('[action-key=true]').click()
    }, 0)
    const res = await dlg.wait()
    expect(res).toBe(true)
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })

  test('Check confirm with handler functions', async () => {
    const dlg = await manager.confirm({
      text: 'test',
      actions: {
        false: 'No',
        true: {
          text: 'Yes',
          class: 'action-true',
          handler: () => {
            return new Promise((resolve) => {
              setTimeout(() => resolve({ msg: 'foo' }), 5)
            })
          }
        }
      }
    })
    expect(dlg.element).toMatchSnapshot()
    setTimeout(() => {
      dlg.element.querySelector('[action-key=true]').click()
      Vue.nextTick(() => {
        expect(dlg.element.querySelector('[action-key=true]').hasAttribute('disabled')).toBe(true)
      })
    }, 0)
    const res = await dlg.wait()
    expect(res).toEqual({ msg: 'foo' })
    await Vue.nextTick()
    expect(document.body.innerHTML).toBe('')
  })
})
