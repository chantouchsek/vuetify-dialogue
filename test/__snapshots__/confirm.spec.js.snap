// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager Check confirm 1`] = `
<div
  class="dialog-layout"
>
  <div
    class="v-card v-sheet theme--light rounded-0"
  >
    <!---->
     
    <div
      class="v-card__text body-1 py-2 pt-4"
    >
      test
    </div>
     
    <div
      class="v-card__actions"
    >
      <div
        class="spacer"
      />
       
      <button
        action-key="ok"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  ok
  
          <!---->
        </span>
      </button>
      <button
        action-key="cancel"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  cancel
  
          <!---->
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`manager Check confirm with action icons 1`] = `
<div
  class="dialog-layout"
>
  <div
    class="v-card v-sheet theme--light rounded-0"
  >
    <!---->
     
    <div
      class="v-card__text body-1 py-2 pt-4"
    >
      test
    </div>
     
    <div
      class="v-card__actions"
    >
      <div
        class="spacer"
      />
       
      <button
        action-key="false"
        class="v-btn v-btn--contained v-btn--rounded theme--light v-size--default info"
        color="info"
        rounded="true"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate material-icons theme--light"
            text="cancel"
          >
            cancel
          </i>
          
  No
  
          <!---->
        </span>
      </button>
      <button
        action-key="true"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  Yes
  
          <i
            aria-hidden="true"
            class="v-icon notranslate v-icon--right material-icons theme--light"
            text="success"
          >
            success
          </i>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`manager Check confirm with btns true|false 1`] = `
<div
  class="dialog-layout"
>
  <div
    class="v-card v-sheet theme--light rounded-0"
  >
    <!---->
     
    <div
      class="v-card__text body-1 py-2 pt-4"
    >
      test
    </div>
     
    <div
      class="v-card__actions"
    >
      <div
        class="spacer"
      />
       
      <button
        action-key="true"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  Yes
  
          <!---->
        </span>
      </button>
      <button
        action-key="false"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  No
  
          <!---->
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`manager Check confirm with handler functions 1`] = `
<div
  class="dialog-layout"
>
  <div
    class="v-card v-sheet theme--light rounded-0"
  >
    <!---->
     
    <div
      class="v-card__text body-1 py-2 pt-4"
    >
      test
    </div>
     
    <div
      class="v-card__actions"
    >
      <div
        class="spacer"
      />
       
      <button
        action-key="false"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  No
  
          <!---->
        </span>
      </button>
      <button
        action-key="true"
        class="v-btn v-btn--contained theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <!---->
          
  Yes
  
          <!---->
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`manager Returnable 1`] = `
<div
  class="dialog-layout"
>
  <p />
</div>
`;

exports[`manager Test confirm without icon 1`] = `
<div
  class="dialog-layout"
>
  <div
    class="v-card v-sheet theme--light rounded-0"
  >
    <header
      class="v-sheet theme--light v-toolbar v-toolbar--dense v-toolbar--flat"
      style="height: 48px;"
    >
      <div
        class="v-toolbar__content"
        style="height: 48px;"
      >
        <!---->
         
        <div
          class="v-toolbar__title"
        >
          
      Title
    
        </div>
      </div>
    </header>
     
    <div
      class="v-card__text body-1 py-2"
    >
      Test confirmation withot icon
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`manager Test default confirm 1`] = `
<div
  class="dialog-layout"
>
  <div
    class="v-card v-sheet theme--light rounded-0"
  >
    <header
      class="v-sheet theme--light v-toolbar v-toolbar--dense v-toolbar--flat"
      style="height: 48px;"
    >
      <div
        class="v-toolbar__content"
        style="height: 48px;"
      >
        <!---->
         
        <div
          class="v-toolbar__title"
        >
          
      Title
    
        </div>
      </div>
    </header>
     
    <div
      class="v-card__text body-1 py-2"
    >
      Test confirmation
    </div>
     
    <!---->
  </div>
</div>
`;
